### PR TITLE
[Exporter.Prometheus] Add allow-utf-8 escaping

### LINKED ISSUE #7246: Support additional Prometheus UTF-8 escaping schemes
Prometheus supports four different metric and label name escaping schemes (see [_UTF-8 metric and label name escaping schemes_](https://prometheus.io/docs/instrumenting/escaping_schemes/)):

- [No Escaping (allow-utf-8)](https://prometheus.io/docs/instrumenting/escaping_schemes/#no-escaping-allow-utf-8)
- [Underscore Escaping (underscores)](https://prometheus.io/docs/instrumenting/escaping_schemes/#underscore-escaping-underscores)
- [Dots Escaping (dots)](https://prometheus.io/docs/instrumenting/escaping_schemes/#dots-escaping-dots)
- [Value Encoding Escaping (values)](https://prometheus.io/docs/instrumenting/escaping_schemes/#value-encoding-escaping-values)

The default behaviour is to use `underscores`, which is what #7209 implements:

> If no escaping scheme is specified in the Accept header, underscores escaping SHOULD be used.

The [default `Accept` header sent by Prometheus](https://prometheus.io/docs/instrumenting/content_negotiation/#default-accept-header) is:

```text
Accept: application/openmetrics-text;version=1.0.0;escaping=allow-utf-8;q=0.5,application/openmetrics-text;version=0.0.1;q=0.4,text/plain;version=1.0.0;escaping=allow-utf-8;q=0.3,text/plain;version=0.0.4;q=0.2,/;q=0.1
```

As shown above, and because it is listed first, `escaping=allow-utf-8` is the preferred scheme to use.

At a minimum we should support `escaping=allow-utf-8` when specified by the client with a fallback to `escaping=underscores`.

### PR DESCRIPTION
**Pull request by martincostello**
Fixes #7246

## Changes

Add support for the `allow-utf-8` escaping scheme.

## Prometheus serializer allocation improvements (this branch vs `gh-7246`)

`PrometheusSerializerBenchmarks` (Prometheus text v1.0.0 / underscores), BenchmarkDotNet Default job + memory diagnoser, .NET 10.0.9, i7-13700H. Both refs benchmarked in the same session (`gh-7246` from a fresh worktree; this branch = allow-utf-8 + label-key allocation optimizations). Allocation normalised to KB per single op.

| Benchmark | Metric | `gh-7246` | This branch | Δ |
|---|---|---:|---:|---:|
| **WriteMetric** | Allocated | 4.12 KB | 3.28 KB | **−20.4%** |
| | Mean (×1) | 2,582 ns | 2,485 ns | −3.7% |
| | Mean (×10000) | 26.30 ms | 25.24 ms | −4.0% |
| **WriteHistogramMetric** | Allocated | 1.26 KB | 1.10 KB | **−12.5%** |
| | Mean (×10000) | 14.41 ms | 14.05 ms | −2.5% |
| **WriteMetricWithTypedLabels** | Allocated | 1.16 KB | 0.73 KB | **−37.3%** |
| | Mean (×10000) | 4.47 ms | 3.90 ms | −12.7% |

Allocation scales linearly across the 1 / 1000 / 10000 call-count params, confirming the per-op figures.

**Summary:** allocation is down **13–37%** in core serialization with neutral-to-faster throughput (StdDevs ~1% this run). The reduction comes from no longer allocating a `StringBuilder` + string for every output label key — already-valid keys are now returned/written with zero allocation. The allow-utf-8 feature itself is allocation-neutral on the common path.

### Detailed Results

<details>

<summary>Expand to view</summary>

#### #7439

```text
BenchmarkDotNet v0.15.8, Windows 11 (10.0.26200.8655/25H2/2025Update/HudsonValley2)
13th Gen Intel Core i7-13700H 2.90GHz, 1 CPU, 20 logical and 14 physical cores
.NET SDK 10.0.301
  [Host]     : .NET 10.0.9 (10.0.9, 10.0.926.27113), X64 RyuJIT x86-64-v3
  DefaultJob : .NET 10.0.9 (10.0.9, 10.0.926.27113), X64 RyuJIT x86-64-v3
```

| Method                     | NumberOfSerializeCalls | Mean            | Error         | StdDev        | Gen0      | Allocated   |
|--------------------------- |----------------------- |----------------:|--------------:|--------------:|----------:|------------:|
| **WriteMetric**                | **1**                      |      **2,581.5 ns** |      **27.57 ns** |      **23.03 ns** |    **0.3357** |     **4.12 KB** |
| WriteHistogramMetric       | 1                      |      1,481.5 ns |      29.18 ns |      28.66 ns |    0.1011 |     1.26 KB |
| WriteMetricWithTypedLabels | 1                      |        441.4 ns |       6.30 ns |       5.89 ns |    0.0939 |     1.16 KB |
| **WriteMetric**                | **1000**                   |  **2,572,046.6 ns** |  **37,995.08 ns** |  **31,727.61 ns** |  **335.9375** |  **4117.19 KB** |
| WriteHistogramMetric       | 1000                   |  1,475,238.5 ns |  25,244.04 ns |  23,613.29 ns |  101.5625 |  1257.81 KB |
| WriteMetricWithTypedLabels | 1000                   |    433,982.6 ns |   4,025.90 ns |   3,143.16 ns |   94.2383 |  1156.25 KB |
| **WriteMetric**                | **10000**                  | **26,298,293.8 ns** | **523,925.09 ns** | **643,427.12 ns** | **3343.7500** | **41171.88 KB** |
| WriteHistogramMetric       | 10000                  | 14,412,322.4 ns | 259,791.33 ns | 243,008.97 ns | 1015.6250 | 12578.13 KB |
| WriteMetricWithTypedLabels | 10000                  |  4,467,305.4 ns |  78,161.99 ns |  69,288.56 ns |  937.5000 |  11562.5 KB |

#### This PR

```text
BenchmarkDotNet v0.15.8, Windows 11 (10.0.26200.8655/25H2/2025Update/HudsonValley2)
13th Gen Intel Core i7-13700H 2.90GHz, 1 CPU, 20 logical and 14 physical cores
.NET SDK 10.0.301
  [Host]     : .NET 10.0.9 (10.0.9, 10.0.926.27113), X64 RyuJIT x86-64-v3
  DefaultJob : .NET 10.0.9 (10.0.9, 10.0.926.27113), X64 RyuJIT x86-64-v3
```

| Method                     | NumberOfSerializeCalls | Mean            | Error         | StdDev        | Gen0      | Allocated  |
|--------------------------- |----------------------- |----------------:|--------------:|--------------:|----------:|-----------:|
| **WriteMetric**                | **1**                      |      **2,485.1 ns** |      **29.39 ns** |      **24.54 ns** |    **0.2670** |     **3360 B** |
| WriteHistogramMetric       | 1                      |      1,417.2 ns |      23.14 ns |      21.65 ns |    0.0896 |     1128 B |
| WriteMetricWithTypedLabels | 1                      |        397.9 ns |       7.65 ns |       6.38 ns |    0.0591 |      744 B |
| **WriteMetric**                | **1000**                   |  **2,519,263.3 ns** |  **43,087.04 ns** |  **40,303.65 ns** |  **265.6250** |  **3360000 B** |
| WriteHistogramMetric       | 1000                   |  1,361,864.6 ns |  17,776.48 ns |  16,628.13 ns |   89.8438 |  1128000 B |
| WriteMetricWithTypedLabels | 1000                   |    391,383.2 ns |   6,740.23 ns |   8,023.77 ns |   59.0820 |   744000 B |
| **WriteMetric**                | **10000**                  | **25,242,336.2 ns** | **329,968.13 ns** | **308,652.39 ns** | **2656.2500** | **33600000 B** |
| WriteHistogramMetric       | 10000                  | 14,048,737.1 ns | 273,508.46 ns | 242,458.12 ns |  890.6250 | 11280000 B |
| WriteMetricWithTypedLabels | 10000                  |  3,901,919.2 ns |  58,909.57 ns |  55,104.05 ns |  589.8438 |  7440000 B |

</details>

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [x] Unit tests added/updated
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [ ] ~~Changes in public API reviewed (if applicable)~~

### CODE REVIEW

File: src/OpenTelemetry.Exporter.Prometheus.AspNetCore/CHANGELOG.md [hunk 1]
```diff
@@ -10,6 +10,14 @@ Notes](../../RELEASENOTES.md).
 * Added a verbose-level diagnostic event for ignored metrics.
   ([#7429](https://github.com/open-telemetry/opentelemetry-dotnet/pull/7429))
 
+* Added support for the `dots` and `values` Prometheus UTF-8 name escaping
+  schemes when negotiated via the `Accept` header.
+  ([#7439](https://github.com/open-telemetry/opentelemetry-dotnet/pull/7439))
+
+* Added support for the `allow-utf-8` Prometheus UTF-8 name escaping scheme
+  when negotiated via the `Accept` header.
+  ([#TODO](https://github.com/open-telemetry/opentelemetry-dotnet/pull/TODO))
```
Comment (martincostello) [inline src/OpenTelemetry.Exporter.Prometheus.AspNetCore/CHANGELOG.md]:
```suggestion
  ([#7440](https://github.com/open-telemetry/opentelemetry-dotnet/pull/7440))
```

File: src/OpenTelemetry.Exporter.Prometheus.HttpListener/CHANGELOG.md [hunk 2]
```diff
@@ -10,6 +10,14 @@ Notes](../../RELEASENOTES.md).
 * Added a verbose-level diagnostic event for ignored metrics.
   ([#7429](https://github.com/open-telemetry/opentelemetry-dotnet/pull/7429))
 
+* Added support for the `dots` and `values` Prometheus UTF-8 name escaping
+  schemes when negotiated via the `Accept` header.
+  ([#7439](https://github.com/open-telemetry/opentelemetry-dotnet/pull/7439))
+
+* Added support for the `allow-utf-8` Prometheus UTF-8 name escaping scheme
+  when negotiated via the `Accept` header.
+  ([#TODO](https://github.com/open-telemetry/opentelemetry-dotnet/pull/TODO))
```
Comment (martincostello) [inline src/OpenTelemetry.Exporter.Prometheus.HttpListener/CHANGELOG.md]:
```suggestion
  ([#7440](https://github.com/open-telemetry/opentelemetry-dotnet/pull/7440))
```

**Comment by Kielek** [comment]
This time from me, not a Codex. Not a bug, it is rather lack of test coverage for utf-8 signs. It will be great to add this. It should be possible, especially through the views.

Not some codex feedback based on this input:

### [P2] Add end-to-end coverage for a multibyte UTF-8 metric name

The `allow-utf-8` serializer snapshots currently use ASCII-only metric names,
such as `http.server.requests` and the name containing quotes, backslashes, and
a line feed. The lower-level escaping and fuzz tests cover Unicode strings, but
they do not exercise a multibyte metric name through `BuildAllowUtf8Names`, the
UTF-8 byte encoding in `NameSet`, and the quoted metadata/sample serialization
paths.

This is a supported scenario: a View-provided metric stream name is not required
to follow the OpenTelemetry instrument name syntax and must not be validated
against it. Please add a deterministic snapshot test that assigns a Unicode name
through a View, for example:

```csharp
new MetricStreamConfiguration
{
    Name = "http.\u82B1\u706B.\U0001F680.requests",
}
```

Ideally, make the test parameterized over all four escaping schemes. These are
name-escaping schemes rather than different wire encodings; the response remains
UTF-8 in every case. The snapshots should verify that:

* `allow-utf-8` preserves the multibyte characters and uses the quoted
  exposition form.
* `underscores` replaces unsupported characters with underscores.
* `dots` encodes dots and replaces unsupported characters.
* `values` reversibly encodes the Unicode code points.

Using a histogram would give especially useful coverage because one test could
verify the quoted or escaped `TYPE`, `HELP`, and `UNIT` metadata names together
with the `_bucket`, `_sum`, `_count`, and `_created` series. Covering both
Prometheus text v1 and OpenMetrics v1 would also exercise the two v1 serializer
implementations.

Specification reference:
https://opentelemetry.io/docs/specs/otel/metrics/sdk/#stream-configuration

**Comment by martincostello** [comment]
> Not a bug, it is rather lack of test coverage for utf-8 signs. It will be great to add this. It should be possible, especially through the views.

Which bits specifically? What did you have in mind to add that isn't already covered or handled by adding some UTF-8 tests suggested by Codex?

**Review by Kielek — APPROVED** [review]
